labelbox_dev/task.py

```diff
@@ -0,0 +1,186 @@
+from functools import lru_cache
+from typing import Any, Dict, Union
+
+import requests
+
+from labelbox_dev.entity import Entity
+from labelbox_dev.exceptions import ResourceNotFoundError
+from labelbox_dev.session import Session
+from labelbox_dev.utils import format_json_to_snake_case
+
+TASK_RESOURCE = "tasks"
+
+logger = logging.getLogger(__name__)
+
+
+class TaskStatus(Enum):
+    COMPLETE = "COMPLETE"
+    FAILED = "FAILED"
+    IN_PROGRESS = "IN_PROGRESS"
+
+
+class TaskType(Enum):
+    DATA_ROW_IMPORT = "DATA_ROW_IMPORT"
+
+
+class BaseTask(Entity, ABC):
+    """ Abstract class representing an asynchronous task
+    """
+
+    task_type: TaskType
+
+    def __init__(self, json):
+        super().__init__(json)
+        self.from_json(json)
+
+    def from_json(self, json) -> "BaseTask":
+        super().from_json(json)
+        self.id = self.json['id']
+        self.updated_at = self.json['updated_at']
+        self.created_at = self.json['created_at']
+        self.status = self.json['status']
+        self.name = self.json['name']
+        self.completion_percentage = self.json['completion_percentage']
+        self.result_url = self.json['result_url']
+
+    @classmethod
+    def get_by_id(cls, task_id) -> "BaseTask":
+        task_json = Session.get_request(f"{TASK_RESOURCE}/{task_id}",
+                                        {'task_type': cls.task_type.name})
+        return cls(task_json)
+
+    def refresh(self) -> "BaseTask":
+        task_json = Session.get_request(f"{TASK_RESOURCE}/{self.id}",
+                                        {'task_type': self.task_type.name})
+        return self.from_json(task_json)
+
+    @abstractmethod
+    def wait_until_done(self):
+        ...
+
+
+class Task(BaseTask, ABC):
+    """ Abstract class representing an asynchronous task with single input.
+    Users can call `task.result` and `task.error` to obtain either the
+    job result upon success, or job error upon failure, respectively.
+    """
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @property
+    @abstractmethod
+    def result(self):
+        ...
+
+    @property
+    @abstractmethod
+    def error(self):
+        ...
+
+
+class BulkTask(BaseTask, ABC):
+    """ Abstract class representing an asynchronous task with bulk input.
+    Users can call `task.results` and `task.errors` to obtain either the
+    job results upon success, or job errors upon failure.
+    """
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @property
+    @abstractmethod
+    def results(self):
+        ...
+
+    @property
+    @abstractmethod
+    def errors(self):
+        ...
+
+
+class DataRowImportTask(BulkTask):
+    """ Asynchronous task representing bulk data_row create job
+    """
+
+    task_type = TaskType.DATA_ROW_IMPORT
+
+    def __init__(self, json):
+        super().__init__(json)
+
+    @lru_cache()
+    def fetch_result(self) -> Dict[str, Any]:
+
+        def download_result():
```

Comment:
what is the reason for the method inside method?

Reply:
No reason, just an artifact of copy pasta. Can remove